### PR TITLE
[FEATURE] Upgrading to DBAL 3 & Switch to symfony/cache

### PR DESCRIPTION
**Pull request by seifane-wise**
### Changes proposed in this pull request:
- Upgrade to DBAL3
- Use symfony/cache for cache since doctrine/cache as been deprecated
- Remove usage from deprecated / deleted classes of DBAL3

I saw too late that a PR was already open concerning this change. I don't know the progress that has been made by @eigan so I opened this one and maybe we can merge our efforts. I think I was able to clear most of the deprecated classes usage in DBAL3. Most significant change is on the cache side where I replaced doctrine/cache by symfony/cache. It broke a lot of tests and I think I managed to fix most of them "correctly". Let me know what you think, if I missed something huge, or if you want to take a different approach on this.

I've been using a version of this branch on my project with no issues so far but I didn't test everything.

Some of the classes I removed (most of them related to codegen) are deprecated on doctrine and will not be replaced by anything. I would argue that if doctrine doesn't want to support codegen anymore it's not the role of this package to implement it. It also keeps the code to a minimum which is easier to in the long run. :) 

Link to the doctrine issue (https://github.com/doctrine/orm/issues/8458)

Link to related PR on laravel-doctrine/migrations (https://github.com/laravel-doctrine/migrations/pull/118)

### CODE REVIEW

**Comment by eigan** [comment]
Could you rebase on laravel-doctrine/orm 1.8. Some of your commits (L9) was cherry-picked and thus have a new sha1. The original commits are now trying to be a part of this PR :) 

EDIT: Some of your commits are duplicates of existing commits, for some reason *

File: src/Configuration/Cache/FileCacheProvider.php [hunk 1]
```diff
@@ -25,13 +25,13 @@ public function __construct(Repository $config)
...
+     * @return CacheItemPoolInterface
      */
     public function resolve(array $settings = [])
     {
-        $path = $settings['path'] ?? $this->config->get('cache.stores.file.path', storage_path('framework/cache'));
+        $path = $settings['namespace'] ?? $this->config->get('doctrine.cache.namespace', 'doctrine-cache');
 
-        return new FilesystemCache(
+        return new FilesystemAdapter(
             $path
```
Comment (eigan) [inline src/Configuration/Cache/FileCacheProvider.php:38]: Third argument is $directory, which I think should be the original $path.

File: src/Configuration/Cache/PhpFileCacheProvider.php [hunk 2]
```diff
@@ -25,14 +25,15 @@ public function __construct(Repository $config)
...
      */
     public function resolve(array $settings = [])
     {
-        $path = $settings['path'] ?? $this->config->get('cache.stores.file.path', storage_path('framework/cache'));
+        $namespace = $settings['namespace'] ?? $this->config->get('doctrine.cache.namespace', 'doctrine-cache');
 
-        return new PhpFileCache(
-            $path
+        return new PhpFilesAdapter(
+            $namespace
```
Comment (eigan) [inline src/Configuration/Cache/PhpFileCacheProvider.php]: Same with third argument here.

File: src/Console/ClearMetadataCacheCommand.php [hunk 3]
```diff
@@ -35,20 +34,16 @@ public function handle(ManagerRegistry $registry)
...
             $em    = $registry->getManager($name);
-            $cache = $em->getConfiguration()->getMetadataCacheImpl();
+            $cache = $em->getConfiguration()->getMetadataCache();
 
             if (!$cache) {
                 throw new InvalidArgumentException('No Result cache driver is configured on given EntityManager.');
             }
 
-            if ($cache instanceof ApcCache) {
+            if ($cache instanceof IlluminateCacheProvider && $cache->getStore() === "apc") {
```
Comment (eigan) [inline src/Console/ClearMetadataCacheCommand.php]: Could check if instanceof ApcCacheProvider? (there is more of these)

**Review by eigan — CHANGES_REQUESTED** [review]

**Comment by eigan** [comment]
@dpslwk Any thoughts? I know you tried to use illuminate/cache (which I would prefer, since it is most likely required by all who use this package), but there was some issues?

**Comment by seifane-wise** [comment]
@eigan I rebased from 1.8 but some they were some conflicts on composer.json that I had to edit. I may have messed up something during rebase that cause this weird commit history. I just committed in response to your review.

Concerning the use of symfony/cache it is a suggested dependency of illuminate/cache if that helps my case :). Also it is recommended by the doctrine team and makes integration easier.

**Comment by eigan** [comment]
@seifane-wise When you rebase, drop the commits not from you. If using `git rebase -i upstream/1.8` can you replace `pick` with `drop` at the commits. Looks like at least three commits(?) dated march 2.

**Comment by seifane-wise** [comment]
@eigan should be good now. commit are correctly rebased.

File: composer.json [hunk 4]
```diff
@@ -18,9 +18,9 @@
     "require": {
         "php": "^8.0",
         "doctrine/annotations": "^1.13",
-        "doctrine/cache": "^1",
-        "doctrine/dbal": "^2.13.3",
-        "doctrine/orm": "^2.11",
+        "doctrine/cache": "^2",
+        "doctrine/dbal": "^3.3",
+        "doctrine/orm": "~2.11.0",
```
Comment (eigan) [inline composer.json]: Why change to ~ here?

File: composer.json [hunk 5]
```diff
@@ -31,7 +31,9 @@
         "illuminate/support": "^9.0",
         "illuminate/validation": "^9.0",
         "illuminate/view": "^9.0",
-        "symfony/serializer": "^6.0"
+        "symfony/serializer": "^6.0",
+        "doctrine/inflector": "^2",
```
Comment (eigan) [inline composer.json:35]: Why was the requirement of inflector ^2 needed?

**Comment by dpslwk** [comment]
not sure about the removal of the YAML in this, does that break the yaml mapping for entities? I thought that we would not need to remove that until orm 3.x, not dabl 3.x

File: composer.json (hunk 4 above)
Comment (seifane-wise) [reply composer.json]: No good reason, it's reverted.

File: composer.json (hunk 5 above)
Comment (seifane-wise) [reply composer.json:35]: It's not needed per se but it was ^1.4 before so I bumped it to ^2

**Comment by seifane-wise** [comment]
sorry @dpslwk I didn't you had plans to remove it in the future. I had a glance over all deprecated class usage and removed as much as I could as I was trying to future proof as much I could right now. Maybe this branch should also be a new version like my PR on migrations ? 2.x or 3.x ?

File: composer.json (hunk 5 above)
Comment (dpslwk) [reply composer.json:35]: we removed it in the 1.8 branch, (see https://github.com/laravel-doctrine/orm/commit/e97d1454cbfd8a5d14dafe59f9b39a7fe6cb72cf) it was ^1.4 in the other branches, so this has added it back in, not bumped it

Comment (seifane-wise) [reply composer.json:35]: Oh my bad I got confused then :/ @dpslwk do you want me to remove it ?

**Comment by SCIF** [comment]
Hey guys,

First of all, am highly appreciate your work!

It's 2.5 months past last activity. Is there any chance to get this stuff done? Moreover, doctrine team [has advised in January](https://www.doctrine-project.org/2022/01/22/sunsetting-dbal-2.html) to migrate to 3.x branch

File: src/Configuration/Connections/MasterSlaveConnection.php [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 namespace LaravelDoctrine\ORM\Configuration\Connections;
 
-use Doctrine\DBAL\Connections\MasterSlaveConnection as MasterSlaveDoctrineWrapper;
+use Doctrine\DBAL\Connections\PrimaryReadReplicaConnection as MasterSlaveDoctrineWrapper;
```
Comment (chewbakartik) [inline src/Configuration/Connections/MasterSlaveConnection.php:5]: How challenging would it be to change the name of the alias as well? If I'm understanding the change that Doctrine underwent, `PrimaryReadReplicaConnection` is intending to replace problematic references in programming to slavery. Obviously this isn't something that would be a functional change, but would be in line with many movements that the software field has been addressing with legacy problematic naming conventions.

File: composer.json (hunk 5 above)
Comment (dpslwk) [reply composer.json:35]: yeah remove it, its include via `doctrine/orm` so no need to call it out

**Comment by dpslwk** [comment]
> sorry @dpslwk I didn't you had plans to remove it in the future. I had a glance over all deprecated class usage and removed as much as I could as I was trying to future proof as much I could right now. Maybe this branch should also be a new version like my PR on migrations ? 2.x or 3.x ?

I really think the YAML meta data driver should be restored, that is deprecated with ORM and will be removed in ORM 3.x its not related to DBAL 3
Yes we will need to address this when ORM 3 is eventually released
Side note my main project that uses ORM and the main reason for my work here currently uses YAML mappings, the longer I can put off having to rewrite 63 mappings the better ;)

**Comment by seifane-wise** [comment]
Sorry I am closing this PR as I dont have the time to maintain it anymore. I will leave my branch open if someone want to base off of it. :pray:

**Comment by eigan** [comment]
@seifane-wise Understandable! Thanks! 🙂

**Pull request closed without merging**